code/game/objects/items/devices/flash/core.dm

```diff
@@ -0,0 +1,235 @@
+		if(percent > 0.75)
+			add_overlay("battery_good")
+		else if(percent > 0.50)
+			add_overlay("battery_low")
+		else if(percent > 0.25)
+			add_overlay("very_low")
+		else
+			add_overlay("battery_none")
+	else
+		add_overlay("battery_hatch")
+
+/obj/item/device/flash/proc/get_external_power_supply()
+	if(isrobot(src.loc))
+		var/mob/living/silicon/robot/R = src.loc
+		return R.cell
+	if(istype(src.loc, /obj/item/rig_module))
+		var/obj/item/rig_module/module = src.loc
+		if(module.holder && module.holder.wearer)
+			var/mob/living/carbon/human/H = module.holder.wearer
+			if(istype(H) && H.back)
+				var/obj/item/weapon/rig/suit = H.back
+				if(istype(suit))
+					return suit.cell
+	return null
+
+/obj/item/device/flash/proc/try_recharge()
+	. = 1
+	if (!power_supply || power_supply.charge >= power_supply.maxcharge || !self_recharge)
+		return 0
+
+	if (use_external_power)
+		var/obj/item/weapon/cell/external = get_external_power_supply()
+		if(!external || !external.use(charge_cost))
+			return 0
+
+	power_supply.give(charge_cost)
+	update_icon()
+
+	addtimer(CALLBACK(src, .proc/try_recharge), recharge_time * 2 SECONDS, TIMER_UNIQUE)
+
+/obj/item/device/flash/attack(mob/living/M as mob, mob/living/user as mob, var/target_zone)
+
+	user.setClickCooldown(DEFAULT_ATTACK_COOLDOWN)
+	user.do_attack_animation(M)
+
+	if(strength_mul <= 0 || !power_supply || !power_supply.checked_use(charge_cost) || !bulb || bulb.is_burnt)
+		playsound(src.loc, 'sound/items/flashlight.ogg', 50, 1)
+		if(M != user)
+			user.visible_message("<span class='notice'>\The [user] sticks \the [src] in \the [src]'s direction, but nothing happens.</span>", \
+				"<span class='notice'>You stick \the [src] in \the [M]'s direction, but nothing happens.</span>", \
+				"<span class='notice'>You hear a click.</span>")
+		else
+			user.visible_message("<span class='notice'>\The [user.name] sticks \the [src] in the air and presses the activation button, but nothing happens.</span>", \
+				"<span class='notice'>You stick \the [src] in the air and press the activation button, but nothing happens.</span>", \
+				"<span class='notice'>You hear a click.</span>")
+		return
+
+	playsound(src.loc, 'sound/weapons/flash.ogg', 50, 1)
+
+	src.update_icon()
+	src.add_overlay("flash_animation")
+
+	if(M == user) //Radial flash
+		user.visible_message("<span class='warning'>\The [user.name] sticks \the [src] in the air and presses the activation button!</span>", \
+			"<span class='warning'>You stick \the [src] in the air and press the activation button!</span>", \
+			"<span class='notice'>You hear a distinct mechanical shutter...</span>")
+		for (var/mob/living/T in view(user))
+			if(T == user)
+				continue
+			var/distance_mod = 1 - min(1,get_dist(user,T)/6)
+			var/total_power = bulb.strength*strength_mul*0.75*distance_mod
+			if(total_power > 0)
+				T.visible_message("<span class='danger'>\The [T] is caught in the flash!</span>", \
```

Comment:
Why exactly is this a visible_message? This'll result in a lot of chat spam if there's a lot of people in range.